### PR TITLE
feat: Add tests for writing RB chunks to Object Store

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #

This is kind of e2e test for the function [load_chunk_to_object_store](https://github.com/influxdata/influxdb_iox/blob/d84f05b2f5eca6a3f95fb7a21f7fe48046231d88/server/src/db.rs#L424) implemented for #1081 to write data to Object Store (OS). In order to test this, a new DB I created then data is loaded in MB, then moved to RB, then loaded to OS.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: data_types/src/chunk.rs [hunk 1]
```diff
@@ -21,8 +21,11 @@ pub enum ChunkStorage {
     /// The chunk is in the Read Buffer (where it can not be mutated)
     ReadBuffer,
 
+    /// The chunk is both in ReadBuffer and Object Store
+    ReadBufferAndObjectStore,
```
Comment (alamb) [inline data_types/src/chunk.rs:25]: 🎉

File: server/Cargo.toml [hunk 2]
```diff
@@ -29,6 +29,7 @@ serde = "1.0"
 serde_json = "1.0"
 snafu = "0.6"
 snap = "1.0.0"
+tempfile = "3.1.0"
```
Comment (alamb) [inline server/Cargo.toml:32]: Since this appears to only be used in tests, it is probably better to add it to the `[dev-dependencies]` section below.

File: server/src/db.rs [hunk 3]
```diff
@@ -895,6 +905,120 @@ mod tests {
...
+        let paths = pq_chunk.object_store_paths();
+        assert_eq!(paths.len(), 1);
+
+        // Check that the path must exist in the object store
+        let path_list = flatten_list_stream(Arc::clone(&object_store), Some(&paths[0]))
+            .await
+            .unwrap();
+        println!("path_list: {:#?}", path_list);
+        assert_eq!(path_list.len(), 1);
+        assert_eq!(path_list, paths.clone());
```
Comment (alamb) [inline server/src/db.rs:986]: 👍

File: server/src/db.rs [hunk 4]
```diff
@@ -895,6 +905,120 @@ mod tests {
...
+
+        // Create a DB given a server id, an object store and a db name
+        let server_id: NonZeroU32 = NonZeroU32::new(10).unwrap();
+        let db_name = "parquet_test_db";
+        let db = Arc::new(make_database(server_id, Arc::clone(&object_store), db_name));
+
+        // Write some line protocols in Mutable buffer of the DB
+        let mut writer = TestLPWriter::default();
+        writer.write_lp_string(db.as_ref(), "cpu bar=1 10").unwrap();
+        writer.write_lp_string(db.as_ref(), "cpu bar=2 20").unwrap();
```
Comment (alamb) [inline server/src/db.rs:949]: Since chunks can have multiple tables in them, it might make sense to write in a line protocol line that has a different measurement (and thus make a new table). Maybe something like:

```rust
        writer.write_lp_string(db.as_ref(), "disk ops=2 20").unwrap();
```

Adding that coverage as a separate test or a follow on PR would also make sense to me

File: server/src/db.rs [hunk 5]
```diff
@@ -895,6 +905,120 @@ mod tests {
...
+            path
+        );
+
+        let mut ctx = context::ExecutionContext::new();
+        let df = ctx.sql(&sql).unwrap();
+        df.collect().await.unwrap();
+
+        // Select data from that table
+        let sql = "SELECT * FROM parquet_table";
+        let content = datafusion_plan_and_collect(&mut ctx, &sql).await.unwrap();
```
Comment (alamb) [inline server/src/db.rs:1009]: FWIW you can probably use `ExecutionContext::sql` again here (rather than `datafusion_plan_and_collect`):

```suggestion
        let content = ctx.sql(&sql).unwrap().collect().await.unwrap();
```

File: server/src/db/chunk.rs [hunk 6]
```diff
@@ -165,9 +166,20 @@ impl DBChunk {
                     estimated_bytes,
                 )
             }
-            Self::ParquetFile { .. } => {
-                unimplemented!("parquet file summary not implemented")
-            }
+            Self::ParquetFile { chunk } => ChunkSummary::new_without_timestamps(
```
Comment (alamb) [inline server/src/db/chunk.rs:169]: 🎉

File: server/src/db/chunk.rs [hunk 7]
```diff
@@ -179,15 +191,15 @@ impl PartitionChunk for DBChunk {
...
-            Self::ParquetFile { .. } => unimplemented!("parquet file not implemented"),
+            Self::ParquetFile { chunk, .. } => chunk.id(),
         }
     }
 
     fn table_summaries(&self) -> Vec<data_types::partition_metadata::TableSummary> {
         match self {
             Self::MutableBuffer { chunk, .. } => chunk.table_summaries(),
             Self::ReadBuffer { chunk, .. } => chunk.table_summaries(),
-            Self::ParquetFile { .. } => unimplemented!("parquet file not implemented"),
```
Comment (alamb) [inline server/src/db/chunk.rs:190]: this whole file is awesome. the Parquet files are finally working!!!

**Review by alamb — APPROVED** [review]
I think this looks great @NGA-TRAN !

File: server/src/db.rs [hunk 8]
```diff
@@ -895,6 +905,120 @@ mod tests {
...
+        let path_list = flatten_list_stream(Arc::clone(&object_store), Some(&paths[0]))
+            .await
+            .unwrap();
+        println!("path_list: {:#?}", path_list);
+        assert_eq!(path_list.len(), 1);
+        assert_eq!(path_list, paths.clone());
+
+        // Get full string path
+        // Todo: it is better if we can get this full path from a function
+        // in object_store::path::Path (will talk with mkm (aka Marko))
```
Comment (mkmik) [inline server/src/db.rs:990]: would `.display()` method on Path work for you?

Comment (mkmik) [reply server/src/db.rs:990]: FWIW, the ObjectStorePath API has been designed with the assumption that we can avoid knowing the string representation of the path; 

This assumption  breaks down since we need to be able to pass it to `CREATE EXTERNAL TABLE parquet_table STORED AS PARQUET LOCATION '{}'`

This is all about encoding guarantees ; `fn display` currently makes very few guarantees, since it's intended just for  displaying and not for passing to other APIs:

```
    /// Like `std::path::Path::display, converts an `ObjectStorePath` to a
    /// `String` suitable for printing; not suitable for sending to
    /// APIs.
```

The core problem here is that on some file systems (e.g. local OS filesystem), paths can contain arbitrary byte sequences that do not form valid utf-8 encoded strings.

This generally not a problem right now since we're the ones generating the path names (we even use percent encoding to ensure that we don't produce "/" or empty strings etc).

But it's important to keep in mind that the domain of valid path names and valid utf-8 strings is not the same and that an encoding scheme is required whenever you move from one to the other.

For example, when passing a file to `CREATE EXTERNAL TABLE parquet_table STORED AS PARQUET LOCATION '{}'`, not only the file has to be utf-8, but it also has to have any single quote (`'`) properly escaped according to the rules of the datafusion flavour of the SQL language (hopefully standard SQL escaping rules).

File: server/src/db.rs [hunk 9]
```diff
@@ -895,6 +905,120 @@ mod tests {
...
+        let root_path = format!("{:?}", root.path());
+        let root_path = root_path.trim_matches('"');
+        let path = format!("{}/{}", root_path, paths[0].display());
+        println!("path: {}", path);
+
+        // Create External table of this parquet file to get its content in a human
+        // readable form
+        let sql = format!(
+            "CREATE EXTERNAL TABLE parquet_table STORED AS PARQUET LOCATION '{}'",
+            path
```
Comment (mkmik) [inline server/src/db.rs:1000]: obligatory: 

image

You need to escape single quotes in the path name.

Fun fact: we percent encode almost all other special characters but we don't percent encode single quotes :-)

Comment (mkmik) [reply server/src/db.rs:1000]: no I was wrong, we do percent encode also single quotes, but we should escape single quotes here anyway since we cannot guarantee that the paths won't have single quotes

File: server/src/db.rs (hunk 8 above)
Comment (alamb) [reply server/src/db.rs:990]: FWIW this `CREATE EXTERNAL TABLE` statement is only used in tests so until we end up testing on machines with 

If we want to avoid using `CREATE EXTERNAL TABLE` on a file, it is possible to register a  `ParquetFile` provider https://docs.rs/datafusion/3.0.0/datafusion/datasource/parquet/struct.ParquetTable.html directly rather than using the `CREATE EXTERNAL TABLE` synctax. At the moment this would require writing to a temp file on local disk as the source can't be in-memory buffer

File: server/src/db.rs (hunk 9 above)
Comment (alamb) [reply server/src/db.rs:1000]: Again, since this is only a test I am not sure how important handing all the edge cases are

Comment (mkmik) [reply server/src/db.rs:1000]: I don't know enough of the context to judge if it's dangerous to have a code snippet in a test that could be copied outside of tests.

Perhaps adding a comment like "we don't care about escaping quotes here because this is just a test"?

File: server/Cargo.toml (hunk 2 above)
Comment (NGA-TRAN) [reply server/Cargo.toml:32]: Ah, something new to learn :)

File: server/src/db.rs (hunk 4 above)
Comment (NGA-TRAN) [reply server/src/db.rs:949]: Yes, I have thought about this after I sent out the PR :) I will add more tests to cover more cases

File: server/src/db.rs (hunk 8 above)
Comment (NGA-TRAN) [reply server/src/db.rs:990]: @mkmik 
> would .display() method on Path work for you?
I use .display() but it only  displays the path we specify using push_dir. The root path before that is not included.

And yes, this is for test only and I think no effort needed for now.

File: server/src/db.rs (hunk 9 above)
Comment (NGA-TRAN) [reply server/src/db.rs:1000]: Good idea. I will add the suggested comments

File: server/src/db.rs (hunk 5 above)
Comment (NGA-TRAN) [reply server/src/db.rs:1009]: Great suggestion for eliminating extra code :)

File: data_types/src/chunk.rs (hunk 1 above)
Comment (crepererum) [reply data_types/src/chunk.rs:25]: How/why is this enum different from this enum? https://github.com/influxdata/influxdb_iox/blob/d84f05b2f5eca6a3f95fb7a21f7fe48046231d88/server/src/db/catalog/chunk.rs#L16-L42

Comment (NGA-TRAN) [reply data_types/src/chunk.rs:25]: This is a very good question and I just discovered this ChunkStorage enum last Friday. I asked the same question. I try to answer here. @alamb: correct me if I am wrong.
ChunkStorage is a mapping of the ChunkState for now but in the future they might not be the same. It mostly tells where we can find a chunk. ChunkState is all about the state of a chunk which I think can be explained similarly for now that is why it is confusing. 

Beside they might be different in the future, I think another difference is ChunkStorage is more lightweight
and easier to use to communicate between objects/functions. Maybe that is why they are defined in proto?

Comment (alamb) [reply data_types/src/chunk.rs:25]: The way I think of it:
* `data_types/src/chunk.rs` is the struct used for sharing the status of a Chunk with clients (it is what the system tables are built from, for example)
* `server/src/db/catalog/chunk.rs` contains the actual data for the Chunk and is internal to the server's `Db`

Does this make sense?

Comment (NGA-TRAN) [reply data_types/src/chunk.rs:25]: Yes, that makes sense
